fix: Update NVD CPE search URLs to match new search interface

chadlwilson opened the pull request:
## Description of Change

NVD appear to have updated their search interface in a way that breaks our earlier generated deep links from reports etc.

The new interface is similar, but the URLs build CPE identifiers rather than individual components. The "applicability" mode appears to be correct (otherwise you get matches where a given vuln is running on/with a CPE but not affecting that CPE?)

Version-specific search
https://nvd.nist.gov/vuln/search#/nvd/home?sortOrder=3&sortDirection=2&cpeFilterMode=applicability&cpeName=cpe:2.3:a:vmware:spring_framework:5.0.0:*:*:*:*:*:*:*&resultType=records

Broad search
https://nvd.nist.gov/vuln/search#/nvd/home?sortOrder=3&sortDirection=2&cpeFilterMode=applicability&cpeName=cpe:2.3:a:vmware:spring_framework:*:*:*:*:*:*:*:*&resultType=records

Open questions 
- [x] should I add a default sort order? Order seems random/database order otherwise.
  - ~~Sort by identifier? descending? ascending?~~
  - Published Date? Descending ✅
- [x] Need to update URLs inside the HTML and Jenkins report templates?
  - These ones go to [URLs like this](https://web.nvd.nist.gov/view/vuln/search-results?adv_search=true&cves=on&cpe_version=cpe%3A%2Fa%3Aghostscript%3Aghostscript%3A8.64) - but they are totally broken/404ed so I am not sure if that recently happened or if they have been broken for a while.

## Related issues

- fixes #7962

## Have test cases been added to cover the new functionality?

yes

Example report generated via Gradle plugin locally:
[dependency-check-report.html](https://github.com/user-attachments/files/22558375/dependency-check-report.html)

chadlwilson committed:
fix: Update NVD CPE search URLs to match new search interface

Signed-off-by: Chad Wilson <29788154+chadlwilson@users.noreply.github.com>

jeremylong commented:
1. We should add a sort order - published date is probably good enough, descending if possible (i.e., newest first).
2. If the links are broken in the HTML report, we likely need to fix them as well. The easiest approach would be to expose the `nvdSearchUrlFor()` methods in the `ReportTool`, as it is already added to the Velocity context.

chadlwilson committed:
fix: Fix bad links in HTML reports for "Vulnerable Software & Versions"

Since the search function accepts a full CPE 2.3 name, we should be able to use whatever CPE we have, with whatever specificity the CPE has. If it happens to have a
version number in it, that will be in the URL. If it does not (due to match being against version ranges, then it will do a broad search).

Signed-off-by: Chad Wilson <29788154+chadlwilson@users.noreply.github.com>

chadlwilson commented:
OK I figured out that a different mechanism was being used for the "Identifiers" vs. the "Vulnerable Software & Versions" sections. Which also had a different URL syntax. Not sure, but suspect the "Vulnerable Software & Versions" links might have been broken for a while.

I've made this change now, however needs a bit more testing.

I *think* we don't need to specifically use the broad URLs, because the search syntax just takes a CPE 2.3 formatted string anyway. If from the NVD match we know details such as the version number from the matched software (sometimes it is included, if the affected version is not a range), it should be fine to include it in the search. In most other cases including `*` should be sufficient.

This new search also allows us to search specficially for all the other pieces of the CPE that have shown up in the NVD match, not just part/vendor/product/version - so behaviour should be better, I think.

chadlwilson commented:
Added an example HTML report generated with this to the description.

chadlwilson committed:
Merge branch 'main' into fix-nvd-search-links
